[PyTorch Edge][type] Add type check in compatibility api

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62612
* #63130
* __->__ #63129
* #65976



1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

Commit:
[PyTorch Edge][type] Add type check in compatibility api

1. Add an api to get `SupportedType` from runtime
2. Expand `is_compatible` api to check type. When checking type, for each type in type table, it will extract all type from a string. For example: `Dict[int, Tuple[Tensor, Tensor, Tensor]]` => [`Dict`, `Tuple`, `Tensor`]. Then compare the list [`Dict`, `Tuple`, `Tensor`] with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime
2. Expand `is_compatible` api to check type. When checking type, for each type in type table, it will extract all type from a string. For example: `Dict[int, Tuple[Tensor, Tensor, Tensor]]` => [`Dict`, `Tuple`, `Tensor`]. Then compare the list [`Dict`, `Tuple`, `Tensor`] with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime
2. Expand `is_compatible` api to check type. When checking type, for each type in type table, it will extract all type from a string. For example: `Dict[int, Tuple[Tensor, Tensor, Tensor]]` => [`Dict`, `Tuple`, `Tensor`]. Then compare the list [`Dict`, `Tuple`, `Tensor`] with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime
2. Expand `is_compatible` api to check type. When checking type, for each type in type table, it will extract all type from a string. For example: `Dict[int, Tuple[Tensor, Tensor, Tensor]]` => [`Dict`, `Tuple`, `Tensor`]. Then compare the list [`Dict`, `Tuple`, `Tensor`] with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime
2. Expand `is_compatible` api to check type. When checking type, for each type in type table, it will extract all type from a string. For example: `Dict[int, Tuple[Tensor, Tensor, Tensor]]` => [`Dict`, `Tuple`, `Tensor`]. Then compare the list [`Dict`, `Tuple`, `Tensor`] with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/mobile/model_compatibility.h:
```diff
@@ -53,6 +53,7 @@ TORCH_API std::unordered_map<std::string, OperatorInfo> _get_model_ops_and_info(
 struct ModelCompatibilityInfo {
   uint64_t bytecode_version;
   std::unordered_map<std::string, OperatorInfo> operator_info;
+  std::vector<at::IValue> type_table;
```
Good question! `std::vector<at::IValue>` is the most accurate way to describe type. Each `IValue` will look like `Dict[Int, Tensor]` and there will many different kinds of combinations. However, it's not easy to compare `std::vector<at::IValue>` with the type list `std::unordered_set<std::string>` from runtime. The current way is just to break down the string and extract the token. I'm still not exactly sure the best layer to do the conversion. The api to get type from model hasn't been implemented yet and I just have the logic to do comparison in `is_compatible` api at this moment.

iseeyuan replied:
I have the same question as @JacobSzwejbka . If there's a type in model of `Dict[Int, Dict[String, custom_class]]`, that is expected to be supported by current code. What would be the type compatibility code work?

cccclai replied:
In latest commit, use `contained_type_` from `type_parser` instead. In the next pr where the custom type is consumed, it should be able to catch this case.

cccclai replied:
With the new implementation (use string instead of Tuple), the following example 
```
"Dict[
   str,
__torch__.A.B.C[
       NamedTuple, [
           [float_features, Tensor],
           [id_list_features, List[Tensor]],
           [label,  Tensor],
           [weight, Tensor],
           ]
       ]
]"
```
will output type set as:
```
(Dict, str, NamedTuple, Tensor, List, Tuple)
```

cccclai commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -220,6 +299,42 @@ ModelCompatCheckResult is_compatible(
     result.errors.emplace_back(s.str());
   }
 
+  SupportedType supported_type = _get_supported_types();
```
Resolved

cccclai commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -179,6 +185,38 @@ std::unordered_map<std::string, OperatorInfo> _get_model_ops_and_info(
   return result;
 }
 
+/********************** Get Type Table **********************/
```
It doesn't require a change in serialization. `type` is an existing field in `bytecode.pkl`. This pr just adds the check, and nothing is changed for serialization/deserialization.

cccclai commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -220,6 +299,42 @@ ModelCompatCheckResult is_compatible(
+      result.status = ModelCompatibilityStatus::ERROR;
+      std::ostringstream s;
+      s << "Primitive type: '" << primitive_type
+        << "' is not support in current runtime";
```
Resolved.

iseeyuan commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -179,6 +185,38 @@ std::unordered_map<std::string, OperatorInfo> _get_model_ops_and_info(
+  std::unordered_set<std::string> all_type_names;
+  for (const auto i : c10::irange(1, bytecode_ivalues.size())) {
+    auto method_tuple = bytecode_ivalues.at(i).toTuple()->elements();
+    auto type_table_tuple = method_tuple.at(1).toTuple()->elements()[3];
```
```suggestion
    auto type_table_tuple = method_tuple.at(1).toTuple()->elements()[BYTECODE_INDEX_TYPE];
```

iseeyuan commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -179,6 +185,38 @@ std::unordered_map<std::string, OperatorInfo> _get_model_ops_and_info(
 
+/********************** Get Type Table **********************/
+
+// Get deduplicate type table given bytecode,
```
Is the return type a vector of IValue of either string or tuple of strings?

iseeyuan commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -203,7 +241,48 @@ ModelCompatibilityInfo ModelCompatibilityInfo::get(
+}
+
+std::unordered_set<std::string> splitString(const std::string& str) {
+  std::unordered_set<std::string> output;
```
Looks like you are implementing a tokenizer

cccclai replied:
Good point. I added a `contained_type` in type parser instead.

iseeyuan commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -203,7 +241,48 @@ ModelCompatibilityInfo ModelCompatibilityInfo::get(
+    possible_primitive_types.insert(
+        primitive_types.begin(), primitive_types.end());
+  }
+  return SupportedType{possible_primitive_types, possible_custom_types};
```
I think the current code accept custom classes with prefix "__torch__.torch.classes". Is it included in this PR?

iseeyuan commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -220,6 +299,42 @@ ModelCompatCheckResult is_compatible(
     result.errors.emplace_back(s.str());
   }
 
+  SupportedType supported_type = runtime_info.supported_types;
```
Consider refactoring this long function to several functions like typeCompactability(), operatorCompactability(), etc.

cccclai replied:
Yeah let me have a separate pr for it

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -31,16 +32,14 @@ class TypeParser {
-    } else if (token == "List") {
+  static std::unordered_set<std::string> getNonSimpleType() {
+    static std::unordered_set<std::string> nonSimpleType{
+        "List", "Optional", "Dict", "Tuple", "__torch__"};
```
`"__torch__"` might be custom class, and not belong to prim types?

iseeyuan replied:
My bad. It's custom class, not custom type. Quite confusing...

iseeyuan reviewed (commented):
Please check my inline comments for details. Major concerns are listed below:
1. For tokenizer and type compatibility check logic. Is there any code that can be shared from type_parser? 
2. Could you add comments on how to check the compatibility of composite type, like `Dict[String, List[classType]]`?

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime
2. Expand `is_compatible` api to check type. When checking type, for each type in type table, it will extract all type from a string. For example: `Dict[int, Tuple[Tensor, Tensor, Tensor]]` => [`Dict`, `Tuple`, `Tensor`]. Then compare the list [`Dict`, `Tuple`, `Tensor`] with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime, including primitive types and custom types.
2. Add a field in `type_parser.cpp` to track the contained types.
3. Expand `is_compatible` api to check type. When checking type, it will check primitive type list and custom type list with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime, including primitive types and custom types.
2. Add a field in `type_parser.cpp` to track the contained types.
3. Expand `is_compatible` api to check type. When checking type, it will check primitive type list and custom type list with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime, including primitive types and custom types.
2. Add a field in `type_parser.cpp` to track the contained types.
3. Expand `is_compatible` api to check type. When checking type, it will check primitive type list and custom type list with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `SupportedType` from runtime, including primitive types and custom types.
2. Add a field in `type_parser.cpp` to track the contained types.
3. Expand `is_compatible` api to check type. When checking type, it will check primitive type list and custom type list with the current support type from current runtime.

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

iseeyuan commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -482,8 +482,23 @@ TEST(LiteInterpreterTest, GetByteCodeVersion) {
+  m._save_for_mobile(ss, {}, true);
+
+  auto contained_types = _get_model_contained_types(ss);
+  AT_ASSERT(contained_types.size() >= 0);
```
Could we test the actual type is expected?

cccclai replied:
I thought about it. I'm a bit hesitate because the type might be changed depending on inline/optimization. Do you think the updated test `test/cpp/jit/test_mobile_type_parser.cpp` in https://github.com/pytorch/pytorch/pull/63130 is sufficient?

iseeyuan commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -179,6 +187,69 @@ std::unordered_map<std::string, OperatorInfo> _get_model_ops_and_info(
+/********************** Get Type Table **********************/
+
+// Forward declare
+std::unordered_set<std::string> _get_model_contained_types(
```
Consider renaming it to `_get_**mobile**_model_contained_types`, because this function only get types from bytecode.

iseeyuan commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -179,6 +187,69 @@ std::unordered_map<std::string, OperatorInfo> _get_model_ops_and_info(
+
+// Get deduplicate type table given bytecode, and each string is a atomic type,
+// like str, Tensor and etc.
+std::unordered_set<std::string> _get_model_contained_types(
```
Could you add an example in the comment, to show the bytecode, and the output of this function?

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -31,16 +38,30 @@ class TypeParser {
-      }
-      return simpleTypeIt->second;
-    } else if (token == "List") {
+  // The list of non-simple types supported by currrent parser.
```
Could you run a size test to record how much size aggression?

cccclai replied:
Yeah for ios,
Compressed : +3.8 KiB (+0.01%) => 73.69 MiB
Uncompressed: +8.7 KiB (+0.01%) => 161.83 MiB

iseeyuan reviewed (commented):
Would be great to have an example of types from a model, as well as a list from a runtime. 
Also could you measure the size regress on type_parser? 
Would the compatibility codes also ship to runtime?

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -160,11 +198,37 @@ class TypeParser {
 }
+
+// Get all contained type given a string
+TORCH_API std::unordered_set<std::string> getContainedTypes(
```
Would those be used in compatibility API only? Could you put them in a different file, since it does not belong to type parsing itself?

cccclai replied:
TypeParser is declared in `type_parser.cpp` only, and other cpp can't access it. Move the class definition to header file is an option. Do you see any issue?

iseeyuan reviewed (commented):
Thanks! The supported type functions may not belong to type_parser.cpp. I'm working on the modularization of runtime. getSupportedType() and related functions may not be needed for all backends. If it's decoupled, this PR is good to go!

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM

Commit:
Update on "[PyTorch Edge][type] Add type check in compatibility api"




1. Add an api to get `supported_types` from runtime, expose in c++ only.
2. Add an api to get `contained_types` from model, expose in both c++ and PyThon.
3. Add a field `contained_types_` in `type_parser.cpp` to track the contained types when parsing python string.
4. Expand `is_compatible` api to check type. When checking type, it will check the contained type list from the model with the support type list from runtime.
5. Expand the unittest for compatibility to cover type
6. Add unit test in python to check type list

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419/)

Differential Revision: [D30231419](https://our.internmc.facebook.com/intern/diff/D30231419)

[ghstack-poisoned]
